src/lsp4clj/json_rpc/messages.clj

```diff
@@ -0,0 +1,63 @@
+(ns lsp4clj.json-rpc.messages)
+
+(set! *warn-on-reflection* true)
+
+(def base-message {:jsonrpc "2.0"})
+
+(defn request
+  ([method params] ;; notification
+   (assoc base-message
+          :method method
+          :params params))
+  ([id method params] ;; request
+   (-> (request method params)
+       (assoc :id id))))
+
+(defn response
+  [id result]
+  (let [response (assoc base-message :id id)]
+    (if (and (map? result) (:error result))
+      (assoc response :error (:error result))
+      (assoc response :result result))))
+
+(def error-codes
+  {;; JSON-RPC errors
+   ;; Is it possible to respond if you have a parse error? How could you reply
+   ;; to a request if you weren't able to parse it?
+   :parse-error      {:code -32700, :message "Parse error"}
```

Comment:
Good point, I always saw that on the spec as: I tried to process a method like `find-references`, but the clojure code had a parse-error, so I could return this, WDYT? 🤷🏻‍♂️

Reply:
True, it's not super easy to tell. The JSON-RPC spec says that there are three levels of error, one for the JSON level, one for the server level (lsp4clj) and one for the application level (clojure-lsp). The spec says that these [specific error codes](https://www.jsonrpc.org/specification#error_object), -32700 and -32600 are for the lowest JSON layer. If the application layer has a parse error, I think it's supposed to return an error >= -32100. I don't really know though—it's not very carefully specified and I doubt anyone relies on the error codes ranges. In any case, I couldn't find a way to return them to the client, though we do return other predefined error codes at times.

Reply:
Yeah, makes sense